File: design-pattern-java/app/src/main/java/design/pattern/java/patterns/behavioral/chain_responsibility/Leaf.java
package design.pattern.java.patterns.behavioral.chain_responsibility;

import java.util.function.Supplier;

public class Leaf implements ComponentWithHandler {
  private Supplier<Boolean> handler;
  private Container container;

  public Leaf(Supplier<Boolean> handler) {
    this.handler = handler;
    this.container = null;
  }

  @Override
  public void handle() {
    if (this.handler != null) {
      this.handler.get();
    } else if (this.container != null) {
      this.container.handle();
    }
  }

  public void setContainer(Container container) {
    this.container = container;
  }
}

File: design-pattern-jvm/app/src/main/kotlin/design/pattern/kotlin/patterns/behavioral/chain_responsibility/Leaf.kt
package design.pattern.kotlin.patterns.behavioral.chain_responsibility

import java.util.function.Supplier

open class Leaf(private val handler: Supplier<Boolean>?) : ComponentWithHandler {
    private var container: Container? =
        null

    override fun handle() {
        if (handler != null) {
            handler.get()
        } else if (container != null) {
            container!!.handle()
        }
    }

    fun setContainer(container: Container?) {
        this.container = container
    }
}